File: ios/Motif/Native/MotifProto.swift
import Foundation

/// Minimal subset of the motif JSON-RPC protocol shapes the native UI
/// needs. Names and field casing match the wire (i.e. the Rust types in
/// motif-proto and motif-server) so we can `JSONDecoder` straight off the
/// wire without custom keys.
enum MotifProto {
    // MARK: - Sessions

    struct SessionInfo: Codable, Identifiable, Hashable, Sendable {
        var name: String
        var workdir: String?
        var created_at: UInt64?
        /// How many WebSocket peers are currently attached to this session.
        /// Optional for forward-compat with motifd builds that predate the
        /// field; treat `nil` as "unknown" rather than "zero".
        var client_count: UInt32?
        var id: String { name }
    }

    /// Peer attached to the same session. Emitted in `session.attach` and
    /// kept in sync by `client.joined` / `client.left` events.
    struct ClientInfo: Codable, Identifiable, Hashable, Sendable {
        var id: String
        var since: UInt64?
    }

    struct SessionListResult: Codable {
        var sessions: [SessionInfo]
    }

    struct SessionCreateParams: Codable {
        var name: String
        // motif-server's CreateParams.workdir is `PathBuf` (non-Optional);
        // sending null here is a deserialize error on the wire. We require
        // a value at the call site instead.
        var workdir: String
    }

    struct SessionCreateResult: Codable {
        var session: SessionInfo
    }

    struct SessionAttachParams: Codable {
        var name: String
        var last_seq: UInt64?
        var term_fg: String?
        var term_bg: String?
    }

    struct SessionAttachResult: Codable {
        var session: SessionInfo
        var client_id: String?
        var clients: [ClientInfo]?
        var ptys: [PtyInfo]?
        var views: [ViewInfo]?
        var active_view: String?
        var last_seq: UInt64?
    }

    struct SessionDestroyParams: Codable {
        var name: String
    }

    // MARK: - PTYs

    enum ShellKind: String, Codable, Sendable {
        case bash, zsh, fish, unknown
    }

    /// All-optional bundle of cheap context the shell's precmd hook
    /// publishes. Old clients ignore unknown fields; new servers tolerate
    /// missing ones — matches Rust's `#[serde(default,
    /// skip_serializing_if = "Option::is_none")]` per field.
    struct ShellContext: Codable, Sendable, Hashable {
        var branch: String?
        var head: String?
        var venv: String?
        var conda: String?
        var node: String?
    }

    struct PtyInfo: Codable, Identifiable, Hashable, Sendable {
        var id: String
        var cmd: String?
        var cwd: String?
        var cols: UInt16
        var rows: UInt16
        var alive: Bool?
        var created_at: UInt64?
    }

    struct PtyListResult: Codable {
        var ptys: [PtyInfo]
    }

    struct PtyCreateParams: Codable {
        var cmd: String?
        var cwd: String?
        var env: [[String]]?    // [[k, v], ...]
        var cols: UInt16
        var rows: UInt16
    }

    struct PtyCreateResult: Codable {
        var info: PtyInfo
    }

    struct PtyWriteParams: Codable {
        var pty_id: String
        /// Wire field is named `data_b64` for backward compat with older
        /// JSON clients; JSONEncoder base64-encodes Data automatically and
        /// MessagePackEncoder writes it as native bin.
        var data: Data
        enum CodingKeys: String, CodingKey {
            case pty_id
            case data = "data_b64"
        }
    }

    struct PtyResizeParams: Codable {
        var pty_id: String
        var cols: UInt16
        var rows: UInt16
    }

    struct PtyKillParams: Codable {
        var pty_id: String
    }

    // MARK: - Views

    /// `view_id == nil` is the wire form of "no active view" — only used
    /// when the last view closes. We always send a real id from the
    /// client side.
    struct ViewActivateParams: Codable {
        var view_id: String?
    }

    struct ViewOpenParams: Codable {
        var spec: ViewSpec
        var activate: Bool
    }

    struct ViewOpenResult: Codable {
        var view: ViewInfo
    }

    struct ViewCloseParams: Codable {
        var view_id: String
    }

    struct ViewMoveParams: Codable {
        var view_id: String
        var to_index: Int
    }

    struct ViewInfo: Codable, Identifiable, Hashable, Sendable {
        var id: String
        var spec: ViewSpec
        var created_at: UInt64?
    }

    /// Tagged enum matching Rust's `#[serde(tag = "kind", rename_all =
    /// "lowercase")]` on `motif_proto::view::ViewSpec`. The wire form is:
    ///
    ///   {"kind":"pty",     "pty_id":"sh-1"}
    ///   {"kind":"preview", "path":"src/main.rs"}
    ///   {"kind":"diff",    "staged":false, "path":"src/main.rs"|null}
    ///   {"kind":"image",   "path":"docs/screenshot.png"}
    ///
    /// `.other` is a forward-compat fallback for any future kind value
    /// — decoders treat it as opaque, and we never re-encode it (the UI
    /// won't try to round-trip an unknown spec).
    enum ViewSpec: Codable, Hashable, Sendable {
        case pty(ptyID: String)
        case preview(path: String)
        case diff(staged: Bool, path: String?)
        case image(path: String)
        case other(typeName: String)

        private enum Keys: String, CodingKey {
            case kind
            case pty_id
            case path
            case staged
        }

        init(from decoder: any Decoder) throws {
            let c = try decoder.container(keyedBy: Keys.self)
            let kind = (try? c.decode(String.self, forKey: .kind)) ?? "?"
            switch kind {
            case "pty":
                let id = (try? c.decode(String.self, forKey: .pty_id)) ?? ""
                self = .pty(ptyID: id)
            case "preview":
                let p = (try? c.decode(String.self, forKey: .path)) ?? ""
                self = .preview(path: p)
            case "diff":
                let staged = (try? c.decode(Bool.self, forKey: .staged)) ?? false
                let path   = try? c.decodeIfPresent(String.self, forKey: .path)
                self = .diff(staged: staged, path: path ?? nil)
            case "image":
                let p = (try? c.decode(String.self, forKey: .path)) ?? ""
                self = .image(path: p)
            default:
                self = .other(typeName: kind)
            }
        }

        func encode(to encoder: any Encoder) throws {
            var c = encoder.container(keyedBy: Keys.self)
            switch self {
            case .pty(let id):
                try c.encode("pty", forKey: .kind)
                try c.encode(id, forKey: .pty_id)
            case .preview(let p):
                try c.encode("preview", forKey: .kind)
                try c.encode(p, forKey: .path)
            case .diff(let staged, let path):
                try c.encode("diff", forKey: .kind)
                try c.encode(staged, forKey: .staged)
                // Match Rust's serde: emit `path` always, null when absent.
                try c.encode(path, forKey: .path)
            case .image(let p):
                try c.encode("image", forKey: .kind)
                try c.encode(p, forKey: .path)
            case .other(let t):
                try c.encode(t, forKey: .kind)
            }
        }
    }

    // MARK: - fs.*

    enum FileType: String, Codable, Sendable {
        case file, dir, symlink
    }

    enum GitFileStatus: String, Codable, Sendable {
        case unmodified, modified, added, deleted, renamed, copied, untracked, ignored, conflicted
    }

    struct TreeEntry: Codable, Sendable, Hashable, Identifiable {
        var name: String
        var type: FileType
        var size: UInt64
        var mtime: UInt64
        var git_status: GitFileStatus?
        var id: String { name }
    }

    struct FsTreeParams: Codable {
        var path: String
        var depth: UInt32?
        var show_hidden: Bool?
    }

    struct FsTreeResult: Codable {
        var path: String
        var entries: [TreeEntry]
    }

    struct FsStatParams: Codable {
        var path: String
    }

    struct FsStatResult: Codable {
        var type: FileType
        var size: UInt64
        var mtime: UInt64
        var git_status: GitFileStatus?
    }

    struct FsReadParams: Codable {
        var path: String
        var max_bytes: UInt64?
    }

    struct FsReadResult: Codable {
        var content_b64: String
        var sha256: String
        var truncated: Bool
        var binary: Bool
        var mime: String?
    }

    /// `expected_sha256 == nil` is the wire form of "I'm creating a new
    /// file / I don't care about racing"; the server still rejects on
    /// `Conflict (-32004)` when it would clobber existing content unless
    /// `force == true`.
    struct FsWriteParams: Codable {
        var path: String
        var content_b64: String
        var expected_sha256: String?
        var force: Bool
    }

    struct FsWriteResult: Codable {
        var sha256: String
    }

    struct FsMkdirParams: Codable {
        var path: String
    }

    struct FsRemoveParams: Codable {
        var path: String
    }

    struct FsRenameParams: Codable {
        var from: String
        var to: String
    }

    // MARK: - git.*

    struct GitFile: Codable, Sendable, Hashable {
        var path: String
        var staged: GitFileStatus
        var unstaged: GitFileStatus
    }

    struct GitStatusParams: Codable {
        var cwd: String?
    }

    struct GitStatusResult: Codable {
        var branch: String?
        var ahead: UInt32
        var behind: UInt32
        var files: [GitFile]
    }

    struct GitDiffParams: Codable {
        var path: String?
        var staged: Bool
        var cwd: String?
    }

    struct GitDiffResult: Codable {
        var patch: String
    }

    /// One row of `git.diffSummary` — same scope as `git.diff` but with
    /// per-file +/- counts pre-computed by the server.
    struct DiffSummaryFile: Codable, Sendable, Hashable {
        var path: String
        var additions: UInt32
        var deletions: UInt32
    }

    struct DiffSummaryResult: Codable {
        var files: [DiffSummaryFile]
    }

    // MARK: - Server-pushed events

    /// Decoded form of a `pty.output` notification's params.
    struct PtyOutputEvent: Decodable {
        var pty_id: String
        /// Wire field name is `data_b64` (kept for compat); JSON decoder
        /// base64-decodes the string, MessagePack decoder reads native bin.
        var data: Data
        var seq: UInt64?
        enum CodingKeys: String, CodingKey {
            case pty_id
            case data = "data_b64"
            case seq
        }
    }

    struct PtyExitedEvent: Decodable {
        var pty_id: String
        var exit_code: Int?
        var seq: UInt64?
    }

    struct PtyResizeEvent: Decodable {
        var pty_id: String
        var cols: UInt16
        var rows: UInt16
        var seq: UInt64?
    }

    struct PtyCwdChangedEvent: Decodable {
        var pty_id: String
        var cwd: String
        var seq: UInt64?
    }

    struct PtyCreatedEvent: Decodable {
        var info: PtyInfo
        var seq: UInt64?
    }

    /// `tree.changed` — fs notify batch. `paths` are absolute paths whose
    /// `fs.tree` listings may now be stale.
    struct TreeChangedEvent: Decodable {
        var paths: [String]
        var seq: UInt64?
    }

    struct GitChangedEvent: Decodable {
        var seq: UInt64?
    }

    struct ClientJoinedEvent: Decodable {
        var client_id: String
        var since: UInt64?
        var seq: UInt64?
    }

    struct ClientLeftEvent: Decodable {
        var client_id: String
        var seq: UInt64?
    }

    struct ViewOpenedEvent: Decodable {
        var view: ViewInfo
        var seq: UInt64?
    }

    struct ViewClosedEvent: Decodable {
        var view_id: String
        var seq: UInt64?
    }

    struct ViewActiveChangedEvent: Decodable {
        var view_id: String?
        var seq: UInt64?
    }

    struct ViewMovedEvent: Decodable {
        var order: [String]
        var seq: UInt64?
    }

    struct PtyShellBootstrappedEvent: Decodable {
        var pty_id: String
        var shell: ShellKind
        var seq: UInt64?
    }

    struct PtyShellContextEvent: Decodable {
        var pty_id: String
        var ctx: ShellContext
        var seq: UInt64?
    }

    /// `pty.command_started` — fires when the shell hands a command line off
    /// to the kernel (FinalTerm 133;C boundary). `text` is the OSC 7770
    /// logical command string, "" when the shell didn't emit one.
    struct PtyCommandStartedEvent: Decodable {
        var pty_id: String
        var block_id: String
        var text: String
        var cwd: String?
        var started_at: UInt64?
        var seq: UInt64?
    }

    struct PtyCommandFinishedEvent: Decodable {
        var pty_id: String
        var block_id: String
        var exit_code: Int?
        var finished_at: UInt64?
        var seq: UInt64?
    }
}

File: apps/flutter/lib/motif/models/motif_proto.dart
/// Wire/domain models for the motif protocol.
///
/// Ported from `apps/ios/Motif/Native/MotifProto.swift`. Field names match the
/// JSON wire format exactly (snake_case) so these decode/encode against the
/// same `motifd` server the iOS app talks to. Keep names in sync with the Rust
/// `motif-proto` crate.
library;

// ─────────────────────────── small helpers ───────────────────────────

int? _asInt(Object? v) => v == null ? null : (v as num).toInt();
String? _asString(Object? v) => v as String?;

// ─────────────────────────── enums ───────────────────────────

enum ShellKind {
  bash,
  zsh,
  fish,
  unknown;

  static ShellKind fromWire(Object? v) => switch (v) {
    'bash' => ShellKind.bash,
    'zsh' => ShellKind.zsh,
    'fish' => ShellKind.fish,
    _ => ShellKind.unknown,
  };

  String get wire => name;
}

enum FileType {
  file,
  dir,
  symlink;

  static FileType fromWire(Object? v) => switch (v) {
    'dir' => FileType.dir,
    'symlink' => FileType.symlink,
    _ => FileType.file,
  };

  String get wire => name;
}

enum GitFileStatus {
  unmodified,
  modified,
  added,
  deleted,
  renamed,
  copied,
  untracked,
  ignored,
  conflicted;

  static GitFileStatus fromWire(Object? v) {
    for (final s in values) {
      if (s.name == v) return s;
    }
    return GitFileStatus.unmodified;
  }

  String get wire => name;
}

// ─────────────────────────── core structs ───────────────────────────

class PingInfo {
  final String service;
  final String version;

  /// LAN-direct hint for a rendezvous server: the plaintext port and this
  /// host's NIC addresses to try at it. Empty/`null` unless the paired motifd
  /// opened a non-loopback `--listen` (see the rzv direct-upgrade path). A
  /// same-LAN client probes these and upgrades off the relay.
  final int? rzvDirectPort;
  final List<String> rzvDirectAddrs;

  const PingInfo({
    required this.service,
    required this.version,
    this.rzvDirectPort,
    this.rzvDirectAddrs = const [],
  });

  factory PingInfo.fromJson(Map<String, Object?> j) => PingInfo(
    service: (j['service'] as String?) ?? '',
    version: (j['version'] as String?) ?? '',
    rzvDirectPort: (j['rzv_direct_port'] as num?)?.toInt(),
    rzvDirectAddrs:
        (j['rzv_direct_addrs'] as List?)?.cast<String>().toList() ?? const [],
  );

  /// True when the peer is a genuine motif-server (vs. some other HTTP service
  /// answering on the port).
  bool get isMotifServer => service == 'motif-server';
}

class SessionInfo {
  final String name;
  final String? workdir;
  final int? createdAt;
  final int? clientCount;

  const SessionInfo({
    required this.name,
    this.workdir,
    this.createdAt,
    this.clientCount,
  });

  String get id => name;

  factory SessionInfo.fromJson(Map<String, Object?> j) => SessionInfo(
    name: (j['name'] as String?) ?? '',
    workdir: _asString(j['workdir']),
    createdAt: _asInt(j['created_at']),
    clientCount: _asInt(j['client_count']),
  );
}

class ClientInfo {
  final String id;
  final int? since;

  const ClientInfo({required this.id, this.since});

  factory ClientInfo.fromJson(Map<String, Object?> j) =>
      ClientInfo(id: (j['id'] as String?) ?? '', since: _asInt(j['since']));
}

class ShellContext {
  final String? branch;
  final String? head;
  final String? venv;
  final String? conda;
  final String? node;

  const ShellContext({
    this.branch,
    this.head,
    this.venv,
    this.conda,
    this.node,
  });

  factory ShellContext.fromJson(Map<String, Object?> j) => ShellContext(
    branch: _asString(j['branch']),
    head: _asString(j['head']),
    venv: _asString(j['venv']),
    conda: _asString(j['conda']),
    node: _asString(j['node']),
  );

  /// Build from the loose string map carried by OSC 777 Context payloads.
  factory ShellContext.fromMap(Map<String, String> m) => ShellContext(
    branch: m['branch'],
    head: m['head'],
    venv: m['venv'],
    conda: m['conda'],
    node: m['node'],
  );
}

class PtyInfo {
  final String id;
  final String? cmd;
  final String? cwd;
  final int cols;
  final int rows;
  final bool? alive;
  final int? createdAt;

  /// Command currently executing, as tracked server-side from shell-integration
  /// markers. Present on cold attach so a restored client can recognize a
  /// running command whose start marker it never saw. Null at a prompt.
  final String? runningCommand;

  const PtyInfo({
    required this.id,
    this.cmd,
    this.cwd,
    required this.cols,
    required this.rows,
    this.alive,
    this.createdAt,
    this.runningCommand,
  });

  factory PtyInfo.fromJson(Map<String, Object?> j) => PtyInfo(
    id: (j['id'] as String?) ?? '',
    cmd: _asString(j['cmd']),
    cwd: _asString(j['cwd']),
    cols: _asInt(j['cols']) ?? 80,
    rows: _asInt(j['rows']) ?? 24,
    alive: j['alive'] as bool?,
    createdAt: _asInt(j['created_at']),
    runningCommand: _asString(j['running_command']),
  );

  PtyInfo copyWith({
    String? cmd,
    String? cwd,
    int? cols,
    int? rows,
    bool? alive,
  }) => PtyInfo(
    id: id,
    cmd: cmd ?? this.cmd,
    cwd: cwd ?? this.cwd,
    cols: cols ?? this.cols,
    rows: rows ?? this.rows,
    alive: alive ?? this.alive,
    createdAt: createdAt,
    runningCommand: runningCommand,
  );
}

/// Tagged union describing what a view renders. Wire form is
/// `{"kind": "...", ...}`.
sealed class ViewSpec {
  const ViewSpec();

  factory ViewSpec.fromJson(Map<String, Object?> j) {
    switch (j['kind']) {
      case 'pty':
        return PtyViewSpec((j['pty_id'] as String?) ?? '');
      case 'preview':
        return PreviewViewSpec((j['path'] as String?) ?? '');
      case 'diff':
        return DiffViewSpec(
          staged: (j['staged'] as bool?) ?? false,
          path: _asString(j['path']),
        );
      case 'image':
        return ImageViewSpec((j['path'] as String?) ?? '');
      default:
        return OtherViewSpec((j['kind'] as String?) ?? 'unknown');
    }
  }

  Map<String, Object?> toJson();
}

class PtyViewSpec extends ViewSpec {
  final String ptyId;
  const PtyViewSpec(this.ptyId);
  @override
  Map<String, Object?> toJson() => {'kind': 'pty', 'pty_id': ptyId};
}

class PreviewViewSpec extends ViewSpec {
  final String path;
  const PreviewViewSpec(this.path);
  @override
  Map<String, Object?> toJson() => {'kind': 'preview', 'path': path};
}

class DiffViewSpec extends ViewSpec {
  final bool staged;
  final String? path;
  const DiffViewSpec({required this.staged, this.path});
  @override
  Map<String, Object?> toJson() => {
    'kind': 'diff',
    'staged': staged,
    if (path != null) 'path': path,
  };
}

class ImageViewSpec extends ViewSpec {
  final String path;
  const ImageViewSpec(this.path);
  @override
  Map<String, Object?> toJson() => {'kind': 'image', 'path': path};
}

class OtherViewSpec extends ViewSpec {
  final String typeName;
  const OtherViewSpec(this.typeName);
  @override
  Map<String, Object?> toJson() => {'kind': typeName};
}

class ViewInfo {
  final String id;
  final ViewSpec spec;
  final int? createdAt;

  const ViewInfo({required this.id, required this.spec, this.createdAt});

  factory ViewInfo.fromJson(Map<String, Object?> j) => ViewInfo(
    id: (j['id'] as String?) ?? '',
    spec: ViewSpec.fromJson((j['spec'] as Map?)?.cast<String, Object?>() ?? {}),
    createdAt: _asInt(j['created_at']),
  );
}

// ─────────────────────────── filesystem / git ───────────────────────────

class TreeEntry {
  final String name;
  final FileType type;
  final int size;
  final int mtime;
  final GitFileStatus? gitStatus;

  const TreeEntry({
    required this.name,
    required this.type,
    required this.size,
    required this.mtime,
    this.gitStatus,
  });

  factory TreeEntry.fromJson(Map<String, Object?> j) => TreeEntry(
    name: (j['name'] as String?) ?? '',
    type: FileType.fromWire(j['type']),
    size: _asInt(j['size']) ?? 0,
    mtime: _asInt(j['mtime']) ?? 0,
    gitStatus: j['git_status'] == null
        ? null
        : GitFileStatus.fromWire(j['git_status']),
  );
}

class FsStatResult {
  final FileType type;
  final int size;
  final int mtime;
  final GitFileStatus? gitStatus;

  const FsStatResult({
    required this.type,
    required this.size,
    required this.mtime,
    this.gitStatus,
  });

  factory FsStatResult.fromJson(Map<String, Object?> j) => FsStatResult(
    type: FileType.fromWire(j['type']),
    size: _asInt(j['size']) ?? 0,
    mtime: _asInt(j['mtime']) ?? 0,
    gitStatus: j['git_status'] == null
        ? null
        : GitFileStatus.fromWire(j['git_status']),
  );
}

class FsReadResult {
  final String contentB64;
  final String sha256;
  final bool truncated;
  final bool binary;
  final String? mime;

  const FsReadResult({
    required this.contentB64,
    required this.sha256,
    required this.truncated,
    required this.binary,
    this.mime,
  });

  factory FsReadResult.fromJson(Map<String, Object?> j) => FsReadResult(
    contentB64: (j['content_b64'] as String?) ?? '',
    sha256: (j['sha256'] as String?) ?? '',
    truncated: (j['truncated'] as bool?) ?? false,
    binary: (j['binary'] as bool?) ?? false,
    mime: _asString(j['mime']),
  );
}

class FsWriteResult {
  final String sha256;
  const FsWriteResult({required this.sha256});
  factory FsWriteResult.fromJson(Map<String, Object?> j) =>
      FsWriteResult(sha256: (j['sha256'] as String?) ?? '');
}

class GitFile {
  final String path;
  final GitFileStatus staged;
  final GitFileStatus unstaged;

  const GitFile({
    required this.path,
    required this.staged,
    required this.unstaged,
  });

  factory GitFile.fromJson(Map<String, Object?> j) => GitFile(
    path: (j['path'] as String?) ?? '',
    staged: GitFileStatus.fromWire(j['staged']),
    unstaged: GitFileStatus.fromWire(j['unstaged']),
  );
}

class GitStatusResult {
  final String? branch;
  final int ahead;
  final int behind;
  final List<GitFile> files;

  const GitStatusResult({
    this.branch,
    required this.ahead,
    required this.behind,
    required this.files,
  });

  factory GitStatusResult.fromJson(Map<String, Object?> j) => GitStatusResult(
    branch: _asString(j['branch']),
    ahead: _asInt(j['ahead']) ?? 0,
    behind: _asInt(j['behind']) ?? 0,
    files: ((j['files'] as List?) ?? [])
        .map((e) => GitFile.fromJson((e as Map).cast<String, Object?>()))
        .toList(),
  );
}

class DiffSummaryFile {
  final String path;
  final int additions;
  final int deletions;

  const DiffSummaryFile({
    required this.path,
    required this.additions,
    required this.deletions,
  });

  factory DiffSummaryFile.fromJson(Map<String, Object?> j) => DiffSummaryFile(
    path: (j['path'] as String?) ?? '',
    additions: _asInt(j['additions']) ?? 0,
    deletions: _asInt(j['deletions']) ?? 0,
  );
}

// ─────────────────────────── attach result ───────────────────────────

class AttachResult {
  final SessionInfo? session;
  final String? clientId;
  final List<ClientInfo> clients;
  final List<PtyInfo> ptys;
  final List<ViewInfo> views;
  final String? activeView;
  final int? lastSeq;
  final String? theme;

  const AttachResult({
    this.session,
    this.clientId,
    this.clients = const [],
    this.ptys = const [],
    this.views = const [],
    this.activeView,
    this.lastSeq,
    this.theme,
  });

  factory AttachResult.fromJson(Map<String, Object?> j) {
    List<T> list<T>(String key, T Function(Map<String, Object?>) f) =>
        ((j[key] as List?) ?? [])
            .map((e) => f((e as Map).cast<String, Object?>()))
            .toList();
    return AttachResult(
      session: j['session'] == null
          ? null
          : SessionInfo.fromJson((j['session'] as Map).cast<String, Object?>()),
      clientId: _asString(j['client_id']),
      clients: list('clients', ClientInfo.fromJson),
      ptys: list('ptys', PtyInfo.fromJson),
      views: list('views', ViewInfo.fromJson),
      activeView: _asString(j['active_view']),
      lastSeq: _asInt(j['last_seq']),
      theme: _asString(j['theme']),
    );
  }
}

class MotifNotification {
  final String title;
  final String body;
  final String? sessionId;
  final String kind;

  const MotifNotification({
    required this.title,
    required this.body,
    this.sessionId,
    required this.kind,
  });

  factory MotifNotification.fromJson(Map<String, Object?> j) =>
      MotifNotification(
        title: (j['title'] as String?) ?? '',
        body: (j['body'] as String?) ?? '',
        sessionId: _asString(j['session_id']),
        kind: (j['kind'] as String?) ?? '',
      );
}
